added ss2 PR test

ambrosejcarr opened the pull request:
This PR contains code to run a test against the SS2 pipeline on PR. It is deterministic and generates a successful result when run in cromwell. 

The notebook will be removed before the branch gets merged; it is just here for testing purposes. 

It is not yet implemented in jenkins, pending some assistance from devops.

rexwangcc commented on test/test_pr_tests.ipynb:
```diff
@@ -2,7 +2,7 @@
   {
    "cell_type": "code",
-   "execution_count": 14,
+   "execution_count": 1,
```
This notebook looks good to me. The only suggestion I could give here is to add workflow labels(such as `"labels":"PR-test"`) to those testing workflows by using the PATCH endpoint, to distinguish them from other workflows.

ambrosejcarr replied:
Got it. I will probably remove the notebook before I merge, as the workflow will be executed with Jenkins. Because I have a few computers that I develop on, I tend to keep the testing code paired with the branches I work on. 

However, I think the repo is cleaner without the notebooks, so will usually remove them unless they contain analysis products (i.e. I'm working on a research ticket) . 

I've added a note to the PR; forgot this time. Sorry!

jsotobroad replied:
i need to learn more about this python notebook stuff :(

jsotobroad commented on test/smartseq2_single_sample/pr/ValidateSmartSeq2SingleCell.wdl:
```diff
@@ -0,0 +1,43 @@
+    docker: "ubuntu:zesty"
+    cpu: 1
+    memory: "3.75 GB"
+    disks: "local-disk 2 HDD"
```
as a random piece of information, with Google you cant ask for less than 10 GB (this wont fail it will just be defaulted to 10)

jsotobroad commented on test/smartseq2_single_sample/pr/ValidateSmartSeq2SingleCell.wdl:
```diff
@@ -0,0 +1,43 @@
+    picard_metrics_hash=$(cat "${alignment_summary_metrics}" | awk 'NF && $1!~/^#/' | md5sum | awk '{print $1}')
+
+    if [ "$counts_hash" != "${expected_counts_hash}" ]; then
+      >&2 echo "counts_hash ($counts_hash) did not match expected hash (${expected_counts_hash})"
```
I'm going back and forth as to whether I would want to see the diff here or not along with the mismatching hashes just to provide a little more debugging information or do we expect someone to dig into the files themselves if their test fails?

ambrosejcarr replied:
Well, one concern is that these files could sometimes be binary, in what case a diff would overload the log. What about the head of a diff? 

I do imagine that people would often need to dig into their files, but wouldn't be opposed to making life easier in cases where a diff solves the problem.

jsotobroad commented on test/smartseq2_single_sample/pr/TestSmartSeq2SingleCellPR.wdl:
```diff
@@ -0,0 +1,59 @@
+  File fastq1
+  File fastq2
+
+  call target.SmartSeq2SingleCell as target {
```
I would be a little weary of aliasing your call with the same alias as your import if nothing else just for readability

ambrosejcarr replied:
Interesting, we do this a lot. How would you fix this? Would `target_wdl` for the import and `target_workflow` for the execution be a good solution?

jsotobroad commented on test/smartseq2_single_sample/pr/ValidateSmartSeq2SingleCell.wdl:
```diff
@@ -0,0 +1,43 @@
+
```
extra `\n`

ambrosejcarr replied:
fixed.

jsotobroad commented on test/smartseq2_single_sample/pr/TestSmartSeq2SingleCellPR.wdl:
```diff
@@ -0,0 +1,59 @@
+import "ss2_single_sample.wdl" as target
+import "ValidateSmartSeq2SingleCell.wdl" as checker
+
+
```
extra `\n` (unless done on purpose)

ambrosejcarr commented:
jenkins retest

ambrosejcarr commented:
@jsotobroad I thought a bit more about the `diff` request. I think it could be difficult to implement since we run against hashes, and don't actively run the `master` pipelines. Thus, we would also need to pass "successful" outputs, which would increase complexity, and I'm not sure where we'd store those files. 

Happy to converse more about this. If you're satisfied with this response, then I think I've addressed all your comments (thanks!).

jsotobroad commented:
That makes a lot of sense to me, thanks for clearing the diff stuff, I aggre.  I really like the `wdl` / `workflow` distinctions in your aliasing 👍  👍

mckinsel commented on test/smartseq2_single_sample/pr/dependencies.json:
```diff
@@ -0,0 +1,9 @@
+  "hisat2.wdl": "/working/pipelines/tasks/hisat2.wdl",
+  "picard.wdl": "/working/pipelines/tasks/picard.wdl",
+  "rsem.wdl": "/working/pipelines/tasks/rsem.wdl"
+}
```
What's going on with these paths? Does the repo get cloned into `/working` when you run the tests in jenkins?

ambrosejcarr replied:
Correct.

ambrosejcarr replied:
Working is also set as `cwd`.

mckinsel replied:
Thanks for clarifying. I think this is probably letting the details of how jenkins is executing the test bleed into the definition of the test. But we can address it when we reorganize the repo.

ambrosejcarr replied:
Sounds good...  I think the long term goal is to run the portability service in lieu of these tests, but would be interested to hear your thoughts during the reorg discussion. :+1:
